### ISSUE TITLE
How to use 0.6.0 with template classes

### ISSUE THREAD

**Issue by ahladik** [IfcOpenShell/IfcOpenShell#1286]
Hello,

using branch 0.6.0, I've written a template class of the type:
`template<class Schema> class myClass { ... }
`
and then I explicitly instantiate the templates with
```
template class myClass<Ifc2x3>; 
template class myClass<Ifc4>;
...
```
If in my template cpp code I insert:
`#include "ifcgeom\IfcGeom.h"`
the compiler emits an error like
`'../ifcparse/IfcSchema-definitions.h': No such file or directory`

This happens because the macro `IfcSchema` is not defined. This doesn't happen in the IfcOpenShell code because IfcSchema is properly defined (like Ifc2x3, Ifc4, ...) at the project properties level and there is a separate project for each of the schemas.

But I wouldn't want to use this approach.
Is there any way to define IfcSchema based on the template instance? Or another workaround of some kind?

**Comment by ahladik** [comment]
Ok, just for the records, here's my solution.

```
#define IfcSchema	Ifc2x3

#include "ifcparse\macros.h"

#define INCLUDE_PARENT_DIR(x) STRINGIFY(ifcparse/x.h)
#include INCLUDE_PARENT_DIR(IfcSchema)
#undef INCLUDE_PARENT_DIR
#define INCLUDE_PARENT_DIR(x) STRINGIFY(ifcparse/x-definitions.h)
#include INCLUDE_PARENT_DIR(IfcSchema)
#undef INCLUDE_PARENT_DIR

#include "myClass.cpp"

template class myClass<IfcSchema>; 
```

**Comment by aothms** [comment]
> at the project properties level and there is a separate project for each of the schemas.
> But I wouldn't want to use this approach.
> Is there any way to define IfcSchema based on the template instance? Or another workaround of some kind?

I think this is a really essential discussion. Ultimately I found that using templates sounds nice in practice, but (a) it's pretty hard to get the small schema variances right (the `#ifdefs from Ifc2x3-definitions.h` is really much easier). (b) error messages and compile time etc.

What was stopping you from using the cmake properties / macro route? Should we work on some sort of CMake scaffolding to enable 3rd party developers to set this up easily? Or some sort of `create-react-app` equivalent.?

**Comment by ahladik** [comment]
Like many of the features of C++, templates provide enough rope to hang oneself.
But in my opinion, in general they allow much better control over types at compile time than a macro.

For the moment, the solution I've laid out looks good to me: I use the IfcOpenShell macros (so the SCHEMA_HAS_xxx are correctly defined) but can still build my template classes, with small .cpp files for explicit instantiation where the IfcSchema definition is obvious and not hidden among the project properties.
And I can still use a single project.

Among the changes I had made was the ability to use both IFC 2x3 and 4 simultaneously. But my code was much less elegant than version 0.6.0. Of course, I'm encountering several difficulties in upgrading given the changes in the code.

There are also issues with the ability to intervene at the level of conversions performed by IfcGeom::Iterator, which I solved at the time by directly modifying the IfcOpenShell code, but which I would much prefer to solve in a more elegant way.

For example, it would be nice to receive callbacks from the conversion functions so that further operations could be performed. I realize I could probably write my own version of IfcGeom::Iterator, but that seems like a tall order.
